tests/integration/ha_tests/application_charm/src/charm.py

```diff
@@ -0,0 +1,209 @@
+
+    @property
+    def _database_config(self):
+        """Returns the database config to use to connect to the MongoDB cluster."""
+        data = list(self.database.fetch_relation_data().values())[0]
+        username, password, endpoints, replset, uris = (
+            data.get("username"),
+            data.get("password"),
+            data.get("endpoints"),
+            data.get("replset"),
+            data.get("uris"),
+        )
+        if None in [username, password, endpoints, replset, uris]:
+            return {}
+
+        return {
+            "user": username,
+            "password": password,
+            "endpoints": endpoints,
+            "replset": replset,
+            "uris": uris,
+        }
+
+    # ==============
+    # Helpers
+    # ==============
+
+    def _start_continuous_writes(self, starting_number: int) -> None:
+        """Start continuous writes to the MongoDB cluster."""
+        if not self._database_config:
+            return
+
+        self._stop_continuous_writes()
+
+        # Run continuous writes in the background
+        proc = subprocess.Popen(
+            [
+                "/usr/bin/python3",
+                "src/continuous_writes.py",
+                self._database_config["uris"],
+                str(starting_number),
+            ]
+        )
+
+        # Store the continuous writes process id in stored state to be able to stop it later
+        self.app_peer_data[PROC_PID_KEY] = str(proc.pid)
+
+    def _stop_continuous_writes(self) -> Optional[int]:
+        """Stop continuous writes to the MongoDB cluster and return the last written value."""
+        if not self._database_config:
+            return None
+
+        if not self.app_peer_data.get(PROC_PID_KEY):
+            return None
+
+        # Send a SIGKILL to the process and wait for the process to exit
+        proc = subprocess.Popen(["pkill", "--signal", "SIGKILL", "-f", "src/continuous_writes.py"])
+        proc.communicate()
+
+        del self.app_peer_data[PROC_PID_KEY]
+
+        # Query and return the max value inserted in the database
+        # (else -1 if unable to query)
+        try:
+            for attempt in Retrying(stop=stop_after_delay(60), wait=wait_fixed(5)):
+                with attempt:
+                    last_written_value = self._max_written_value()
+        except RetryError as e:
+            logger.exception("Unable to query the database", exc_info=e)
+            return -1
+
+        return last_written_value
+
+    def _max_written_value(self) -> int:
+        """Returns the count of rows in the continuous writes table."""
+        if not self._database_config:
+            return -1
+
+        client = MongoClient(self._database_config["uris"])
+        db = client[DATABASE_NAME]
+        test_collection = db["test_collection"]
+        last_written_value = test_collection.find_one(sort=[("number", -1)])
+        client.close()
+        return last_written_value["number"]
```

Comment:
The last written value should be taken from `src/continuous_writes.py` (not from the database).
For example, `src/continuous_writes.py` can think that it is written 100500 records successfully, but only 500 were written.
Keep in mind the tests below rely on this output.